File: firmware/adc.py
from machine import Pin, SPI

class KalmanFilter:
    def __init__(self, initial_estimate, process_variance, measurement_variance):
        self.estimate = initial_estimate
        self.estimate_error = process_variance

        self.process_variance = process_variance
        self.measurement_variance = measurement_variance

    def update(self, measurement):
        prediction = self.estimate
        prediction_error = self.estimate_error + self.process_variance

        kalman_gain = prediction_error / (prediction_error + self.measurement_variance)
        self.estimate = prediction + kalman_gain * (measurement - prediction)
        self.estimate_error = (1 - kalman_gain) * prediction_error

        return self.estimate


class MCP3008:
    def __init__(self, spi, cs, ref_voltage=3.3):
        self.cs = cs
        self.cs.value(1)
        self._spi = spi
        self._out_buf = bytearray(3)
        self._out_buf[0] = 0x01
        self._in_buf = bytearray(3)
        self._ref_voltage = ref_voltage

    def read(self, pin, is_differential=False):
        self.cs.value(0)
        self._out_buf[1] = ((not is_differential) << 7) | (pin << 4)
        self._spi.write_readinto(self._out_buf, self._in_buf)
        self.cs.value(1)
        return ((self._in_buf[1] & 0x03) << 8) | self._in_buf[2]


class ADC:
    def __init__(self, sck, mosi, miso, cs, baudrate=100000):
        cs = Pin(cs, Pin.OUT)
        cs.value(1)
        self.adc = MCP3008(SPI(0, sck=Pin(sck), mosi=Pin(mosi), miso=Pin(miso), baudrate=baudrate), cs)

    
    def read(self, channel):
        return self.adc.read(channel)

File: firmware/adc.cpp
#include "adc.h"

#include <Adafruit_MCP3008.h>

#define WINDOW_SIZE 75

ADC::ADC(uint8_t sck, uint8_t mosi, uint8_t miso, uint8_t cs, boolean c5,
         boolean c6, boolean c7)
    : adc(), sck(sck), mosi(mosi), miso(miso), cs(cs), c5(c5), c6(c6), c7(c7) {}

void ADC::setup() {
  adc.begin(sck, mosi, miso, cs);

  values = new float *[8];
  lastValues = new float[8];

  for (int i = 0; i < 8; i++) {
    values[i] = new float[WINDOW_SIZE];
    for (int j = 0; j < WINDOW_SIZE; j++) {
      values[i][j] = (i <= 4 || c5 && i == 5 || c6 && i == 6 || c7 && i == 7)
                         ? (float)adc.readADC(i)
                         : 0;
    }

    lastValues[i] = values[i][0];
  }
}

void ADC::update() {
  for (int c = 0; c < 8; c++) {
    if (!c5 && c == 5)
      continue;
    if (!c6 && c == 6)
      continue;
    if (!c7 && c == 7)
      continue;

    float oldValue = values[c][WINDOW_SIZE - 1];
    float newValue = adc.readADC(c);

    for (int i = WINDOW_SIZE - 1; i > 0; i--) {
      values[c][i] = values[c][i - 1];
    }

    values[c][0] = newValue;

    float valuePercentage = toPercentage(values[c][0]);
    float oldValuePercentage = toPercentage(oldValue);

    if (abs(lastValues[c] - valuePercentage) >= 0.01) {
      Serial.print("Channel ");
      Serial.print(c);
      Serial.print(": ");
      Serial.print(String(oldValuePercentage, 2));
      Serial.print(" -> ");
      Serial.println(String(valuePercentage, 2));

      lastValues[c] = valuePercentage;
    }
  }
}

float ADC::toPercentage(float value) { return value / 1023; }